state-sync/diem-data-client/src/diemnet/tests.rs

```diff
@@ -182,6 +203,171 @@ async fn test_request_works_only_when_data_available() {
+
+    let good_peer = mock_network.add_connected_peer();
+    let bad_peer = mock_network.add_connected_peer();
+
+    // bypass poller and just add the storage summaries directly
+
+    // good peer advertises txns 0 -> 100
+    client.update_summary(good_peer, mock_storage_summary(100));
+    // bad peer advertises txns 0 -> 200 (but can't actually service)
+    client.update_summary(bad_peer, mock_storage_summary(200));
+    client.update_global_summary_cache();
+
+    // the global summary should contain the bad peer's advertisement
+    let global_summary = client.get_global_data_summary();
+    assert!(global_summary
+        .advertised_data
+        .transactions
+        .contains(&CompleteDataRange::new(0, 200).unwrap()));
+
+    // spawn a handler for both peers
+    tokio::spawn(async move {
+        while let Some((peer, _, _, response_sender)) = mock_network.next_request().await {
+            if peer == good_peer.peer_id() {
+                // good peer responds with good response
+                response_sender.send(Ok(StorageServiceResponse::TransactionsWithProof(
+                    TransactionListWithProof::new_empty(),
+                )));
+            } else if peer == bad_peer.peer_id() {
+                // bad peer responds with error
+                response_sender.send(Err(StorageServiceError::InternalError("".to_string())));
+            }
+        }
+    });
+
+    let mut seen_data_unavailable_err = false;
+
+    // sending a bunch of requests to the bad peer's upper range will fail
+    for _ in 0..20 {
+        let result = client
+            .get_transactions_with_proof(200, 200, 200, false)
+            .await;
+
+        // while the score is still decreasing, we should see a bunch of
+        // InternalError's. once we see a `DataIsUnavailable` error, we should
+        // only see that error.
+        if !seen_data_unavailable_err {
+            assert_err!(&result);
+            if let Err(Error::DataIsUnavailable(_)) = result {
+                seen_data_unavailable_err = true;
+            }
+        } else {
+            assert_matches!(result, Err(Error::DataIsUnavailable(_)));
+        }
+    }
+
+    // peer should eventually get ignored and we should consider this request
+    // range unserviceable.
+    assert!(seen_data_unavailable_err);
+
+    // the global summary should no longer contain the bad peer's advertisement
+    client.update_global_summary_cache();
+    let global_summary = client.get_global_data_summary();
+    assert!(!global_summary
+        .advertised_data
+        .transactions
+        .contains(&CompleteDataRange::new(0, 200).unwrap()));
+
+    // we should still be able to send the good peer a request
+    let response = client
+        .get_transactions_with_proof(100, 50, 100, false)
+        .await
+        .unwrap();
+    assert_eq!(response.payload, TransactionListWithProof::new_empty());
+}
+
+#[tokio::test]
+async fn bad_peer_is_eventually_banned_callback() {
```

Comment:
Future nit: Another test that would be good to add is one that ensures the peer gets added back once it executes enough get data summary requests.